### ISSUE TITLE
Optimizers memory usage

### ISSUE THREAD

**Issue by jeremiedb** [apache/incubator-mxnet#10928]
There's been few comments about memory consumption in the R-package. With large models, the memory footprint can rapidly burst the available memory unless a manual gc() is added in the training loop. 

The problematic appears to be the weights update performed by the optimizers: https://github.com/apache/incubator-mxnet/blob/master/R-package/R/model.R#L221

Memory behaves as if the temporary nd.arrays were kept in memory. 

Looked at how the python optimizers were designed and took a first attempt at using the `mx.nd.sgd.update` type of operators. 
It seemed to help, but looking at https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer.py#L434, I confused at how the state of optimizers such as momentum or adam and cie is updated. I can't see where the staste is updated other than for the initialization. 

Also, tried to do a direct mutation of the weight using the out parameters of the `mx.nd.sgd.update` to point to the executor ref.arg.arrays, but it got the following message after a first update: 

```
 Error in mx.nd.sgd.update(weight = weight, out = out, grad = grad, lr = learning.rate,  : 
  ./ndarray.h:87: RCheck failed: ptr_->writable && !ptr_->moved Passing a read only NDArray to mutate function 
```
I'm still missing some pieces to figure out a proper way of fixing the memory consumtion of the optimizers. Any help would be much welcome!

@thirdwing

**Comment by vdantu** [comment]
@sandeep-krishnamurthy : Could you please tag this as "question", "R"

@jeremiedb : You could also post this on "discuss.mxnet.io". You would get a wider audience.

**Comment by jeremiedb** [comment]
As an additional pointer, the issue seems tied to R's NDArray being set to read only after an initial mutation. https://github.com/apache/incubator-mxnet/blob/master/R-package/src/ndarray.h#L86 @tqchen ? 

Basic comparative example between R and Python: 

R: 
```
> state <- mx.nd.array(c(1,2,3,3))
> grad <- mx.nd.array(1:4)
> weight <- mx.nd.array(1:4)
> mx.nd.sgd.mom.update(weight, grad, state, lr = 0.1, momentum = 0.5, out = weight)
[1] 1.4 2.8 4.2 5.1
> mx.nd.sgd.mom.update(weight, grad, state, lr = 0.1, momentum = 0.5, out = weight)
Error in mx.nd.sgd.mom.update(weight, grad, state, lr = 0.1, momentum = 0.5,  : 
  ./ndarray.h:87: RCheck failed: ptr_->writable && !ptr_->moved Passing a read only NDArray to mutate function

```

Python:

```
state  = mx.nd.array([1, 2, 3, 3])
grad  = mx.nd.array([1,2,3,4])
weight = mx.nd.array([1,2,3,4])
print(nd.sgd_mom_update(weight, grad, state, momentum = 0.5, lr=0.1, out=weight))
print(nd.sgd_mom_update(weight, grad, state, momentum = 0.5, lr=0.1, out=weight))

[1.4 2.8 4.2 5.1]
<NDArray 4 @cpu(0)>

[1.5  3.   4.5  5.25]
<NDArray 4 @cpu(0)>
```

**Comment by nswamy** [comment]
closing issue as the referenced PR seems to be resolving it, feel free to open a new issue if you still find problems.
